src/transformers/modeling_utils.py

```diff
@@ -1581,22 +1585,33 @@ def _autoset_attn_implementation(
                 "sdpa",
                 "flash_attention_2",
             ]:
-                message = f'Specified `attn_implementation="{config._attn_implementation}"` is not supported. The only possible arguments are `attn_implementation="eager"` (manual attention implementation)'
+                message = (
+                    f'Specified `attn_implementation="{config._attn_implementation}"` is not supported. The only '
+                    'possible arguments are `attn_implementation="eager"` (manual attention implementation)'
+                )
                 if cls._supports_flash_attn_2:
                     message += ', `"attn_implementation=flash_attention_2"` (implementation using flash attention 2)'
                 if cls._supports_sdpa:
-                    message += ', `"attn_implementation=sdpa"` (implementation using torch.nn.functional.scaled_dot_product_attention)'
+                    message += (
+                        ', `"attn_implementation=sdpa"` '
+                        '(implementation using torch.nn.functional.scaled_dot_product_attention)'
+                    )
                 raise ValueError(message + ".")
 
-            # If a config is passed with a preset attn_implementation, we skip the automatic dispatch and use the user-provided config, with hard checks that the requested attention implementation is available.
+            # If a config is passed with a preset attn_implementation, we skip the automatic dispatch and use the
+            # user-provided config, with hard checks that the requested attention implementation is available.
             requested_attn_implementation = config._attn_implementation_internal
+        else:
+            config._attn_implementation_autoset = True
```

Comment:
The actual change (1): if the user DOESN'T set `attn_implementation`, then it was automatically inferred

Reply:
It is needed to preserve the functionality of `_attn_implementation_autoset` -- this flag tells us whether `_attn_implementation` is defined by the user or inferred by our code.

Consider the following examples:

1 - No `attn_implementation`
```py
model = Idefics2ForConditionalGeneration.from_pretrained(
    "HuggingFaceM4/idefics2-8b-base",
    torch_dtype=torch.bfloat16,
    device_map="auto",
)
```

2 - `attn_implementation` set by the user
```py
model = Idefics2ForConditionalGeneration.from_pretrained(
    "HuggingFaceM4/idefics2-8b-base",
    torch_dtype=torch.bfloat16,
    attn_implementation="eager",
    device_map="auto",
)
```

Prior to these changes, `_attn_implementation_autoset ` would always be `True`, which defeats the point of the flag. After these changes, it will only be `True` if the user doesn't fully specify the `attn_implementation` at load time 🤗

Reply:
Oh I see, I guess it was a misunderstanding. The flag was actually introduced to mark that we had already run `autoset_attn_impl` for a given config so we don't have to run it several times. 

There was a weird bug in composite models that if we call the method twice the `requested_attn_impl` is changed to SDPA from `None`, thus raising errors when it shouldn't

Reply:
Oh! Okay, in that case let me rename your flag (`_attn_implementation_autoset_was_run`), and create a new one (`_attn_implementation_was_set_by_autoset `)

Reply:
> I guess either way is good but to make sure, so if user requests some specific attn and we don't support it, will that raise an error or not?

We should raise an error, as otherwise it would be a dark pattern -- the user has requested X, but we are doing Y

Reply:
Yes, we should raise an error saying which module failed to dispatch, perfect!

And I think we don't need both flags as long as the weird bug is not raised in current PR. It was the Blip-2 model failing on some tests when `attn_impl=None`. Btw, this flag is also unset back to False when saving, so if we change name to be more explicit, we've to change it there also